### ISSUE TITLE
False Positives on disallowedFunctionCalls Rule in Version 4.1.1

### ISSUE THREAD

**Issue by yoannblot** [spaze/phpstan-disallowed-calls#288]
# Description
## Summary:
After upgrading from version `4.0.1` to `4.1.1`, the `disallowedFunctionCalls` rule produces false positives when certain strings are present in method names, incorrectly identifying them as forbidden function calls.

## Steps to Reproduce:

Configure `disallowedFunctionCalls` as follows:
```yaml
parameters:
  disallowedFunctionCalls:
    - function:
        - 'data_get()'
        - 'data_set()'
        - 'optional()'
      message: 'This function is not allowed'
```
Define a method or function with a name that includes the substring of a forbidden function, such as `testItHandlesSomeOptional()`.
Run `PHPStan` analysis.

## Observed Behavior:
PHPStan flags the method as violating the `disallowedFunctionCalls` rule:
```yaml
message: '#^Calling optional\(\) is forbidden, This function is not allowed\.$#'
```

## Expected Behavior:
`PHPStan` should not flag method or function names that merely include substrings matching forbidden function names. Only actual calls to the forbidden functions should trigger a violation.

## Environment:

PHPStan version: 2.1.1
spaze/phpstan-disallowed-calls: 4.1.1
Previous working version: 4.0.1

## Additional Notes:

This appears to be a regression introduced in version `4.1.1`. The pattern matching for the `disallowedFunctionCalls` rule seems to be overly aggressive, treating method or function names as if they were calls to the forbidden functions.

Let me know if more details are needed!

**Comment by spaze** [comment]
Hi, this case is actually tested for quite some time already:

`printf` is disallowed:
https://github.com/spaze/phpstan-disallowed-calls/blob/c921da2a5d743e2ed16ba9002a457be3b757319f/tests/Calls/FunctionCallsTest.php#L49-L53

Then in the tested file there's also `printfunk` on line 17:
https://github.com/spaze/phpstan-disallowed-calls/blob/c921da2a5d743e2ed16ba9002a457be3b757319f/tests/src/disallowed/functionCalls.php#L17

But there's no error reported on line 17:
https://github.com/spaze/phpstan-disallowed-calls/blob/c921da2a5d743e2ed16ba9002a457be3b757319f/tests/Calls/FunctionCallsTest.php#L240-L247

4.1.0 (and 4.1.1) has started detecting disallowed calls also in callables (e.g. `array_map('disallowedFunction', [])`) and in some other places, please check [the release notes](https://github.com/spaze/phpstan-disallowed-calls/releases/tag/v4.1.1). Can you please check whether that's not a forbidden function being called that way? Or can you provide a working testfile I can run? You can start with modifying the files above and then run the test with

```
vendor/bin/phpunit tests/Calls/FunctionCallsTest.php
```

**Comment by yoannblot** [comment]
Thanks for your answer 🙏 

I've made a minimal failing test, and for an unknown reason, it only fails with a disallowed function called `optional` 🤔 

# Failing test
## Configuration

```yaml
includes:
  - ./quality/phpstan/vendor/spaze/phpstan-disallowed-calls/extension.neon

parameters:
  level: 5

  disallowedFunctionCalls:
    - function:
        - 'optional()'
```

## Code to analyze
```php
use PHPUnit\Framework\TestCase;

final class NothingTest extends TestCase
{
    public function testItDoesNothing(): void
    {
        $this->assertSame('optional', $this->do());
    }

    private function do(): string
    {
        return 'optional';
    }
}
```

## Result
<img width="473" alt="Image" src="https://github.com/user-attachments/assets/1b3b5adb-c089-46c1-8dae-c5846eb66192" />

# Success test

## Code
I success my analyzing by switching the `assertSame()` with an `assertEquals()` : 
```php
$this->assertEquals('optional', $this->do());
```

## Result
<img width="463" alt="Image" src="https://github.com/user-attachments/assets/ae63562b-2b91-4372-a4e0-929dfce6d0b0" />


That's a bit weird, and seems an edge case, anyway it could hide a more specific bug 🐛 , wdyt?

**Comment by spaze** [comment]
PHPStan 2.x says that the "expected" param of the `assertSame` method is callable, see the test here:

```
string(15) "param->getName:"
string(8) "expected"
string(18) "param->isCallable:"
string(3) "Yes"
```

https://github.com/spaze/phpstan-disallowed-calls/actions/runs/12911818495/job/36005135640#step:6:19
(It's the https://github.com/spaze/phpstan-disallowed-calls/tree/spaze/callable-param branch)

And because `optional()` is most probably an existing function in your code, why would you like to disallow it if it wasn't 😅, then the string `optional` is treated as a callable reference to that function. If the function wouldn't exist, then this condition here would be false, and that's probably why only `optional` triggers this behavior in your case (my test uses `foobar`):

https://github.com/spaze/phpstan-disallowed-calls/blob/c921da2a5d743e2ed16ba9002a457be3b757319f/src/RuleErrors/DisallowedCallableParameterRuleErrors.php#L90-L92

And `optional` is reported because your configuration disallows it.

If the param wouldn't be reported as callable, then this condition here would be false and no error would be reported:

https://github.com/spaze/phpstan-disallowed-calls/blob/c921da2a5d743e2ed16ba9002a457be3b757319f/src/RuleErrors/DisallowedCallableParameterRuleErrors.php#L178-L180

PHPStan 1.12 doesn't report the parameter to be callable:
```
string(15) "param->getName:"
string(8) "expected"
string(18) "param->isCallable:"
string(5) "Maybe"
```
https://github.com/spaze/phpstan-disallowed-calls/actions/runs/12911818495/job/36005136087#step:6:19

Honestly, I don't know what needs to be changed here, if anything in this extension.

**Comment by yoannblot** [comment]
Ok, I see, so there is something weird between `PHPStan 2.x` & `PHPUnit`...
Thank you so much @spaze  for your investigation 🙏 

FYI, I am using `Laravel` framework, which contains too many ways to write ugly code, that's why I want to disallow existing methods like `optional()` ([link](https://laravel.com/docs/11.x/helpers#method-optional)).

**Comment by yoannblot** [comment]
Not a bug here 👍

**Comment by spaze** [comment]
Maybe @ondrejmirtes could shed some light on why PHPStan 2.x shows `PHPUnit\Framework\TestCase::assertSame()`'s param `expected` as callable = yes, while PHPStan 1.x shows it as callable = maybe. It's still possible that I'm doing something funky somewhere 😁

This is the test subject (`spaze/callable-param` branch):

https://github.com/spaze/phpstan-disallowed-calls/blob/941f6ee029ce2b932bdb6bae798cbc08376e6ba6/tests/src/disallowed/testCase.php

and the testcase here:

https://github.com/spaze/phpstan-disallowed-calls/blob/941f6ee029ce2b932bdb6bae798cbc08376e6ba6/tests/Calls/MethodCallsTest.php

Running `vendor/bin/phpunit tests/Calls/MethodCallsTest.php` after `composer update` shows failure (and some intentional vardumps), while after `composer update --prefer-lowest phpstan/phpstan phpstan/phpstan-deprecation-rules --with-all-dependencies` shows Ok (but risky because of vardumps).

**Comment by ondrejmirtes** [comment]
I know! You're trying to get the signatures of PHPUnit's `assertSame` parameters, right?

This is how the signature looks like:

```php
    /**
     * @template ExpectedType
     * @param ExpectedType $expected
     * @throws ExpectationFailedException
     * @phpstan-assert =ExpectedType $actual
     */
    final public static function assertSame(mixed $expected, mixed $actual, string $message = ''): void
```

You're doing:

```php
$parametersAcceptor = ParametersAcceptorSelector::selectFromArgs($scope, $node->getArgs(), $reflection->getVariants());
```

Which triggers the whole generics type inference. What you get back for this call `$this->assertSame('foobar', 'whatever')` is:

```php
    /**
     * @param 'foobar' $expected
     * @throws ExpectationFailedException
     * @phpstan-assert ='foobar' $actual
     */
    final public static function assertSame(mixed $expected, mixed $actual, string $message = ''): void
```

So when you ask for the type of `$expected`, you get `'foobar'` because ConstantStringType knows there's a function of the same name!

In PHPStan 1.x (without bleeding edge) you'd just get `string` instead of `'foobar'` which is just maybe-callable.

As a possible fix you should just get `$reflection->getVariants()` and skip if `count !== 1`. The get the only one with `[0]`. You can still call ArgumentsNormalizer with that.

**Comment by spaze** [comment]
Wow, excellent! Thanks @ondrejmirtes for the analysis and the proposed fix, it works and doesn't report errors in the `assertSame` call.

**Comment by spaze** [comment]
I've released the fix in 4.2.1 now, @yoannblot please let me know if it works as expected now.

**Comment by yoannblot** [comment]
> I've released the fix in 4.2.1 now, [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/7677012?u=2d4cc5232990104f6c4aa31daef20b12d9b114f1&amp;v=4&amp;size=80">@yoannblot](https://github.com/yoannblot?rgh-link-date=2025-01-23T18%3A05%3A26.000Z) please let me know if it works as expected now.

It works like a charm! Thank you so much @spaze & @ondrejmirtes ❤ !
